File: app/src/main/java/com/akshayltc/chime/adapter/FileBrowserAdapter.java
package com.akshayltc.chime.adapter;

import android.app.Activity;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.akshayltc.chime.R;
import com.akshayltc.chime.callback.BaseListener;
import com.akshayltc.chime.utilities.BaseEvents;

import java.io.File;
import java.io.FileFilter;
import java.util.Arrays;
import java.util.Comparator;

public class FileBrowserAdapter extends RecyclerView.Adapter {

    private File currentDir;
    private File baseDir;
    private File[] subdirList;
    private File[] fileList;
    private int subDirCount;
    private int fileCount;
    private Context context;
    private BaseListener mBaseListener;

    public FileBrowserAdapter (File baseDir, Context context, BaseListener mBaseListener) {
        this.baseDir = baseDir;
        this.context = context;
        this.mBaseListener = mBaseListener;
        updateDir(baseDir);
    }

    @NonNull
    @Override
    public FileViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        return new FileViewHolder(LayoutInflater.from(context).inflate(R.layout.file_browser_directory_item, viewGroup, false));
    }

    @Override
    public void onBindViewHolder(@NonNull final RecyclerView.ViewHolder viewHolder, int i) {
        ((FileViewHolder) viewHolder).position = i;
        if ( i == 0 && !currentDir.getName().equals( baseDir.getName() )) {
            ((FileViewHolder) viewHolder).name.setText("..");
            ((FileViewHolder) viewHolder).name.setTextColor(ContextCompat.getColor(context, R.color.secondary_text));
            ((FileViewHolder) viewHolder).RLFile.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    updateDir(currentDir.getParentFile());
                }
            });
        } else if ( currentDir.getName().equals( baseDir.getName() )  ) {
            if ( i < subDirCount ) {
                ((FileViewHolder) viewHolder).name.setText(subdirList[i].getName());
                ((FileViewHolder) viewHolder).name.setTextColor(ContextCompat.getColor(context, R.color.secondary_text));
                final File fileItem = subdirList[((FileViewHolder) viewHolder).position];
                ((FileViewHolder) viewHolder).RLFile.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        if (fileItem.isDirectory()) {
                            updateDir(fileItem);
                        } else {
                            Toast.makeText(context, "Not a Directory", Toast.LENGTH_SHORT).show();
                        }
                    }
                });
            } else {
                ((FileViewHolder) viewHolder).name.setText(fileList[i-subDirCount].getName());
                ((FileViewHolder) viewHolder).name.setTextColor(ContextCompat.getColor(context, R.color.cyan_text));
                final File fileItem = fileList[((FileViewHolder) viewHolder).position - subDirCount];
                ((FileViewHolder) viewHolder).RLFile.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        if (fileItem.isDirectory()) {
                            updateDir(fileItem);
                        } else {
                            Toast.makeText(context, "Not a Directory", Toast.LENGTH_SHORT).show();
                        }
                    }
                });
            }
        } else if ( i != 0 && !currentDir.getName().equals( baseDir.getName() )  ) {
            if ( i <= subDirCount ) {
                ((FileViewHolder) viewHolder).name.setText(subdirList[i - 1].getName());
                ((FileViewHolder) viewHolder).name.setTextColor(ContextCompat.getColor(context, R.color.secondary_text));
                final File fileItem = subdirList[((FileViewHolder) viewHolder).position - 1];
                ((FileViewHolder) viewHolder).RLFile.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        if (fileItem.isDirectory()) {
                            updateDir(fileItem);
                        } else {
                            Toast.makeText(context, "Not a Directory", Toast.LENGTH_SHORT).show();
                        }
                    }
                });
            } else {
                   ((FileViewHolder) viewHolder).name.setText(fileList[i - (subDirCount + 1)].getName());
                ((FileViewHolder) viewHolder).name.setTextColor(ContextCompat.getColor(context, R.color.cyan_text));
                final File fileItem = fileList[((FileViewHolder) viewHolder).position - (subDirCount + 1)];
                ((FileViewHolder) viewHolder).RLFile.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        if (fileItem.isDirectory()) {
                            updateDir(fileItem);
                        } else {
                            Toast.makeText(context, "Not a Directory", Toast.LENGTH_SHORT).show();
                        }
                    }
                });
            }
        }
    }

    public void updateDir(File fileItem) {
        Toast.makeText(context, "Directory", Toast.LENGTH_SHORT).show();
        currentDir = fileItem;
        getFiles();
        notifyDataSetChanged();
        mBaseListener.onEvent( BaseEvents.DIRECTORY_CHANGED, 0, currentDir);
    }

    @Override
    public int getItemCount() {
        if ( currentDir.getName().equals( baseDir.getName() ) ) {
            return subdirList.length + fileList.length;
        } else {
            return subdirList.length + fileList.length + 1;
        }
    }

    public class FileViewHolder extends RecyclerView.ViewHolder {
        TextView name;
        RelativeLayout RLFile;
        int position;

        public FileViewHolder(View itemView) {
            super(itemView);
            // get the reference of item view's
            name = itemView.findViewById(R.id.fileName);
            RLFile = itemView.findViewById(R.id.RLFile);

        }
    }

    public void getFiles() {
        subdirList = currentDir.listFiles(new FileFilter() {
            @Override
            public boolean accept(File file) {
                return file.isDirectory();
            }
        });
        subDirCount = subdirList.length;

        fileList = currentDir.listFiles(new FileFilter() {
            @Override
            public boolean accept(File file) {
                return file.getPath().endsWith(".mp3")
                        || file.getPath().endsWith(".flac") || file.getPath().endsWith(".wav")
                        || file.getPath().endsWith(".ogg");
            }
        });
        fileCount = fileList.length;
    }

    public void sortByName(File[] fileset) {
        Arrays.sort(fileset, new Comparator<File>() {
            @Override
            public int compare(File f1, File f2) {
                if ((f1 == null) || (f2 == null))
                {
                    return 0;
                }

                return 0;
            }
        });
    }

}

File: app/src/main/java/com/whitelotusapps/chime/adapter/FileBrowserAdapter.kt
package com.whitelotusapps.chime.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.RelativeLayout
import android.widget.TextView
import android.widget.Toast
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.whitelotusapps.chime.R
import com.whitelotusapps.chime.callback.BaseListener
import com.whitelotusapps.chime.utilities.BaseEvents
import java.io.File
import java.util.*

class FileBrowserAdapter(private val baseDir: File, private val context: Context, private val mBaseListener: BaseListener) : RecyclerView.Adapter<RecyclerView.ViewHolder?>() {
    private var currentDir: File? = null
    private var subdirList: Array<File>? = null
    private var fileList: Array<File>? = null
    private var subDirCount = 0
    private var fileCount = 0
    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): FileViewHolder {
        return FileViewHolder(LayoutInflater.from(context).inflate(R.layout.file_browser_directory_item, viewGroup, false))
    }

    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, i: Int) {
        (viewHolder as FileViewHolder).pos = i
        if (i == 0 && currentDir!!.name != baseDir.name) {
            viewHolder.name.text = ".."
            viewHolder.name.setTextColor(ContextCompat.getColor(context, R.color.secondary_text))
            viewHolder.RLFile.setOnClickListener { updateDir(currentDir!!.parentFile) }
        } else if (currentDir!!.name == baseDir.name) {
            if (i < subDirCount) {
                viewHolder.name.text = subdirList!![i].name
                viewHolder.name.setTextColor(ContextCompat.getColor(context, R.color.secondary_text))
                val fileItem = subdirList!![viewHolder.position]
                viewHolder.RLFile.setOnClickListener {
                    if (fileItem.isDirectory) {
                        updateDir(fileItem)
                    } else {
                        Toast.makeText(context, "Not a Directory", Toast.LENGTH_SHORT).show()
                    }
                }
            } else {
                viewHolder.name.text = fileList!![i - subDirCount].name
                viewHolder.name.setTextColor(ContextCompat.getColor(context, R.color.cyan_text))
                val fileItem = fileList!![viewHolder.position - subDirCount]
                viewHolder.RLFile.setOnClickListener {
                    if (fileItem.isDirectory) {
                        updateDir(fileItem)
                    } else {
                        Toast.makeText(context, "Not a Directory", Toast.LENGTH_SHORT).show()
                    }
                }
            }
        } else if (i != 0 && currentDir!!.name != baseDir.name) {
            if (i <= subDirCount) {
                viewHolder.name.text = subdirList!![i - 1].name
                viewHolder.name.setTextColor(ContextCompat.getColor(context, R.color.secondary_text))
                val fileItem = subdirList!![viewHolder.position - 1]
                viewHolder.RLFile.setOnClickListener {
                    if (fileItem.isDirectory) {
                        updateDir(fileItem)
                    } else {
                        Toast.makeText(context, "Not a Directory", Toast.LENGTH_SHORT).show()
                    }
                }
            } else {
                viewHolder.name.text = fileList!![i - (subDirCount + 1)].name
                viewHolder.name.setTextColor(ContextCompat.getColor(context, R.color.cyan_text))
                val fileItem = fileList!![viewHolder.position - (subDirCount + 1)]
                viewHolder.RLFile.setOnClickListener {
                    if (fileItem.isDirectory) {
                        updateDir(fileItem)
                    } else {
                        Toast.makeText(context, "Not a Directory", Toast.LENGTH_SHORT).show()
                    }
                }
            }
        }
    }

    fun updateDir(fileItem: File?) {
        Toast.makeText(context, "Directory", Toast.LENGTH_SHORT).show()
        currentDir = fileItem
        files
        notifyDataSetChanged()
        mBaseListener.onEvent(BaseEvents.DIRECTORY_CHANGED, 0, currentDir)
    }

    override fun getItemCount(): Int {
        return if (currentDir!!.name == baseDir.name) {
            subDirCount + fileCount
        } else {
            subDirCount + fileCount
        }
    }

    inner class FileViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var name: TextView
        var RLFile: RelativeLayout
        var pos = 0

        init {
            // get the reference of item views
            name = itemView.findViewById(R.id.fileName)
            RLFile = itemView.findViewById(R.id.RLFile)
        }
    }

    val files: Unit
        get() {
            subdirList = currentDir!!.listFiles { file -> file.isDirectory }
            subDirCount = if (subdirList != null) subdirList!!.size else 0
            fileList = currentDir!!.listFiles { file ->
                (file.path.endsWith(".mp3")
                        || file.path.endsWith(".flac") || file.path.endsWith(".wav")
                        || file.path.endsWith(".ogg"))
            }
            fileCount = if (fileList != null) fileList!!.size else 0
        }

    fun sortByName(fileset: Array<File>?) {
        Arrays.sort(fileset) { f1, f2 ->
            if (f1 == null || f2 == null) {
                0
            } else 0
        }
    }

    init {
        updateDir(baseDir)
    }
}